lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1406,12 +1406,17 @@ class DifferentiableActivityInfo {
   /// by the `return` instruction.
   SmallVector<SILValue, 4> outputValues;
 
+  /// The set of varied variables, indexed by the corresponding independent
+  /// value (input) index.
+  SmallVector<SmallDenseSet<SILValue>, 4> variedValueSets;
   /// The set of useful variables, indexed by the corresponding dependent value
   /// (output) index.
   SmallVector<SmallDenseSet<SILValue>, 4> usefulValueSets;
-  /// The set of useful variables, indexed by the corresponding independent
-  /// value (input) index.
-  SmallVector<SmallDenseSet<SILValue>, 4> variedValueSets;
+
+  /// The set of instructions that have propagated variedness, indexed by the
+  /// corresponding independent value (input) index.
+  SmallVector<SmallDenseSet<SILInstruction *>, 4>
```

Comment:
> This data is used to prevent the algorithm from revisiting instructions that it has already marked as varied. Can you achieve the same thing in a simpler way, without storing this data at all, by making `setVariedAndPropagateToUsers` bail out early when the value is already set as varied?

Your first suggestion worked! Done in https://github.com/apple/swift/pull/28155/commits/2a0e377133dff15209fe9dbc4923366051a52bc6, from commit description:

```
    - Remove `propagatedVariedInstructionSets` and related utilities.
      - Instead of keeping track of instructions that have propagated variedness,
        `setVariedAndPropagateToUsers` now skips already-varied values. This is
        simpler and achieves the same effect.
```